store/sqlstore/user_store.go

```diff
@@ -456,119 +457,59 @@ func applyRoleFilter(query sq.SelectBuilder, role string, isPostgreSQL bool) sq.
-		schemeGuest := false
-		schemeAdmin := false
-		schemeUser := false
+func applyMultiRoleFilters(query sq.SelectBuilder, roles []string, teamRoles []string, channelRoles []string, isPostgreSQL bool) sq.SelectBuilder {
+	sqOr := sq.Or{}
 
+	if len(roles) > 0 && roles[0] != "" {
 		for _, role := range roles {
+			queryRole := wildcardSearchTerm(role)
 			switch role {
-			case model.SYSTEM_ADMIN_ROLE_ID:
-				schemeAdmin = true
 			case model.SYSTEM_USER_ROLE_ID:
-				schemeUser = true
-			case model.SYSTEM_GUEST_ROLE_ID:
-				schemeGuest = true
-			}
-		}
-
-		if schemeAdmin || schemeUser || schemeGuest {
-			if schemeAdmin && schemeUser {
-				queryString += `(u.Roles LIKE '%system_user%' OR u.Roles LIKE '%system_admin%') `
-			} else if schemeAdmin {
-				queryString += `(u.Roles LIKE '%system_admin%') `
-			} else if schemeUser {
-				queryString += `(u.Roles LIKE '%system_user%' AND u.Roles NOT LIKE '%system_admin%') `
-			}
-
-			if schemeGuest {
-				if queryString != "" {
-					queryString += "OR "
+				// If querying for a `system_user` ensure that the only role they have is `system_user`.
+				sqOr = append(sqOr, sq.Eq{"u.Roles": role})
+			case model.SYSTEM_GUEST_ROLE_ID, model.SYSTEM_ADMIN_ROLE_ID, model.SYSTEM_USER_MANAGER_ROLE_ID, model.SYSTEM_READ_ONLY_ADMIN_ROLE_ID, model.SYSTEM_MANAGER_ROLE_ID:
+				// If querying for any other roles search using a wildcard.
+				if isPostgreSQL {
+					sqOr = append(sqOr, sq.ILike{"u.Roles": queryRole})
+				} else {
+					sqOr = append(sqOr, sq.Like{"u.Roles": queryRole})
 				}
-				queryString += `(u.Roles LIKE '%system_guest%') `
 			}
+
 		}
 	}
 
 	if len(channelRoles) > 0 && channelRoles[0] != "" {
-		schemeGuest := false
-		schemeAdmin := false
-		schemeUser := false
 		for _, channelRole := range channelRoles {
 			switch channelRole {
 			case model.CHANNEL_ADMIN_ROLE_ID:
-				schemeAdmin = true
+				sqOr = append(sqOr, sq.And{sq.Eq{"cm.SchemeAdmin": true}, sq.Eq{"u.Roles": model.SYSTEM_USER_ROLE_ID}})
 			case model.CHANNEL_USER_ROLE_ID:
-				schemeUser = true
+				sqOr = append(sqOr, sq.And{sq.Eq{"cm.SchemeUser": true}, sq.Eq{"cm.SchemeAdmin": false}, sq.Eq{"u.Roles": model.SYSTEM_USER_ROLE_ID}})
 			case model.CHANNEL_GUEST_ROLE_ID:
-				schemeGuest = true
-			}
-		}
-
-		if schemeAdmin || schemeUser || schemeGuest {
-			if queryString != "" {
-				queryString += "OR "
-			}
-			if schemeAdmin && schemeUser {
-				queryString += `(cm.SchemeUser = true AND u.Roles = 'system_user')`
-			} else if schemeAdmin {
-				queryString += `(cm.SchemeAdmin = true AND u.Roles = 'system_user')`
-			} else if schemeUser {
-				queryString += `(cm.SchemeUser = true AND cm.SchemeAdmin = false AND u.Roles = 'system_user')`
-			}
-
-			if schemeGuest {
-				if queryString != "" && queryString[len(queryString)-3:] != "OR " {
-					queryString += "OR "
-				}
-				queryString += `(cm.SchemeGuest = true AND u.Roles = 'system_guest')`
+				sqOr = append(sqOr, sq.Eq{"cm.SchemeGuest": true})
```

Comment:
Why do some of these cases not need to check for the `Roles` value?

Reply:
The logic here is basically designed to only get a user if their "highest level" role matches the queried role so if the query is for a `team_admin` then we want to exclude users that are `system_admins` and only grab `system_users`

But for team / channel guest we dont need to do that since a guest is a guest in all scenarios